test/e2e/storage/csi_mock_volume.go

```diff
@@ -761,7 +754,16 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
 					{expectedMethod: "NodeUnstageVolume", expectedError: codes.OK},
 				},
 				// Fail first 3 NodeStage requests, 4th succeeds
-				nodeStageScript: `console.log("Counter:", ++counter); if (counter < 4) { INVALIDARGUMENT; } else { OK; }`,
+				nodeStageHook: func() mockservice.NodeStageHook {
+					counter := 0
+					return func(context.Context, *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error) {
+						counter++
+						if counter < 4 {
+							return nil, status.Error(codes.InvalidArgument, "fake error")
+						}
+						return nil, nil
+					}
+				}(),
```

Comment:
The Go callbacks are obviously longer than the one-line JS code. That's partly because Go is more verbose (strict typing), partly because the callbacks are more flexible (more parameters).

Personally I prefer writing everything in one language with compile-time checks, but perhaps it would be worthwhile to support also simpler callbacks?

One idea that I had was to turn each hook into a struct that accepts a full-blown callback with hook-specific parameters and a generic one that just takes `ctx context.Context, counter int`, with the counter maintained by code in the mock driver. That code then can also deal with concurrency better - in this code, access to `counter` is not protected against concurrent calls (a theoretic issue, but still not quite correct).

Would that be worthwhile?

Reply:
IMO, more verbose hooks are expected in golang and OK, the example above looks quite readable. And every hook can put a lock around their counter (or use atomic int) if they want. I don't see a big issue here.

Reply:
I've pushed a revised implementation where the hooks are implemented entirely inside the e2e.test binary in a generic way via a gRPC interceptor. The API towards the mock driver then becomes "replace normal interceptor with custom one", which is something that we can add to csi-test and then vendor into k/k. For now we still need a fork until we have a csi-test release without JavaScript dependency.

Inside the mock tests, hooks also became simpler thanks to a small helper function which provides the common code for counting method calls.

If we need more complex functionality, then more complex hooks can inspect the request and produce a response in addition to errors.